Rails3 Members Autocompletion with Select2.

GarstgerUnhold opened the pull request:
This is a Rails3 port of https://github.com/opf/openproject/pull/22.

GarstgerUnhold committed:
get multiple remote selection to work properly

GarstgerUnhold committed:
removes dead code formatting divs we don't want to format

Conflicts:
	public/stylesheets/application.css

GarstgerUnhold committed:
adjusting focus color for select2 container

Conflicts:
	public/stylesheets/application.css
	public/stylesheets/select2_customizing.css

Conflicts:
	public/stylesheets/application.css
	public/stylesheets/select2_customizing.css

GarstgerUnhold committed:
starting on infinite scrolling of select2 inside members form. member itself might be working, roles are definitely not

Conflicts:

	app/controllers/members_controller.rb
	app/models/principal.rb

Conflicts:

	app/controllers/members_controller.rb

removes deprecated test for autocompletion

Conflicts:

	Gemfile
	app/views/members/_new_member_form.rhtml
	public/javascripts/members_select_boxes.js

Conflicts:
	Gemfile

Conflicts:
	Gemfile
	test/functional/members_controller_test.rb

tessi commented on Gemfile.lock:
```diff
@@ -373,6 +407,7 @@ DEPENDENCIES
   rails-dev-tweaks (~> 0.6.1)
   rails-footnotes (>= 3.7.5.rc4)
   rails_autolink
+  rb-fsevent
```
please do not commit the rb-fsevent dependency in Gemfile.lock. It is a mac-only dependency and we agreed to not commit it (to prevent "commit fights" between linux and mac users)

ulferts commented:
Haven't looked through the code to much until now. Using the application I noticed the following:
- There is a translation missing for when no role is selected but the form is submitted (bug exists prior to this pull request).
- Role selection seems to be case sensitive. Inserting "role" will not find the role "Role 001".
- There is no feedback (ajax spinner) when actually adding members by clicking "Add". This is also true when deleting members.
- Editing members in order to change their roles is impossible. Clicking "Edit" does nothing except deleting the displayed roles.
- It is impossible to search for first and last name. Inserting "John Doe" will not find the user "John Doe".
- When inserting something and removing the focus (pressing ESC) while the autocompleter loads will lead to the ajax spinner never disappearing.
- Layout of the "Add"-Button is defective. It jumps around depending on which autocomplete field is longer. Should keep to the left.
- It is hard to find roles because you typically do not know the name of a role. I propose to load all roles upfront, maybe limit it to x, so the user can browse through the list.
- The comleter (at least when loading the search results) is vulnerable to XSS injection.
- The accessibility fallback does not work. The search query is always caught in the find_model_object_and_project before filter. 
- There are absolutely no tests. Granted, was true before this pull request. But it would be good to have at least some integration tests.

GarstgerUnhold committed:
port to rails3

Conflicts:
	Gemfile.lock

GarstgerUnhold commented:
Can't seem to reproduce this issue:
- Role selection seems to be case sensitive. Inserting "role" will not find the role "Role 001".

Your "Role 001" wasn't for some reason builtin (e.g. builtin=1 in db)?

GarstgerUnhold commented:
- There is no feedback (ajax spinner) when actually adding members by clicking "Add". This is also true when deleting members.

This is probably due to jQuerys ajax transport being something else then Prototypes XMLHttpRequest (as of 1.5.x, I think), so our old Prototype Responders will not attach properly. It'll require a more general solution outside of this PR.

GarstgerUnhold committed:
allow Users to be scoped with full name, too

GarstgerUnhold commented:
- When inserting something and removing the focus (pressing ESC) while the autocompleter loads will lead to the ajax spinner never disappearing.

This was a bug in select2. I'll update Select2, but we should switch to the select2-rails gem (once it is updated to 3.3.2.), see https://github.com/opf/openproject/pull/60. This is the line that's missing in our select2: https://github.com/ivaynberg/select2/blob/3.3.2/select2.js#L1434

GarstgerUnhold committed:
start loading roles right away

GarstgerUnhold committed:
escape principal and role names in members autocompleter

GarstgerUnhold committed:
fix non-impaired members tab submit button: used to jump around because of floating siblings, applied 'clearfix'

GarstgerUnhold committed:
fixes partial name for html request in memberscontroller's autocompletion

GarstgerUnhold commented:
I guess those commits cover the last of @ulferts pointers. The remaining bugs/features won't be adressed in here, as they don't concern this code.

apfelfabrik commented:
This cannot be merged w/o conflicts, could you merge the branch you are pushing against once more, or rebase?

GarstgerUnhold committed:
Fixes Members Creation:

The Select2 queries return comma-seperated elements that we need to handle
individually. The way we seperate those elements was broken, resulting in a
failing test.

GarstgerUnhold commented:
Did a little rebasin' and fixed a failing test (3b90822). There are still cuces failing, tho.

GarstgerUnhold committed:
fix id seperation used inside autocompletion for impaired users

GarstgerUnhold committed:
renamed user and role id input for non-impaired users, gives us prettier generated ids

GarstgerUnhold commented:
Cuces should not be failing anymore. They are a little brittle if focus is taken from the browser, though. Will rebase the whole thing again.

GarstgerUnhold committed:
autocompletion for roles and members in project settings, now with infinite scrolling

Conflicts:

	Rakefile
	app/controllers/members_controller.rb
	app/models/principal.rb
	app/models/role.rb
	app/views/layouts/base.rhtml
	app/views/members/_new_member_form.rhtml
	public/javascripts/members_select_boxes.js
	public/stylesheets/select2_customizing.css

Conflicts:

	app/views/layouts/base.rhtml
	public/javascripts/members_select_boxes.js

don't need to mark members form as 'new'. everyone will see cause it's awesome

Conflicts:

	app/views/projects/settings/_members.rhtml

Conflicts:

	app/views/members/_new_member_form.rhtml

Conflicts:

	app/views/projects/settings/_members.rhtml

Conflicts:

	app/views/members/_member_form_non_impaired.rhtml
	app/views/projects/settings/_members.rhtml

Conflicts:

	app/controllers/members_controller.rb
	app/views/layouts/base.rhtml
	app/views/members/_autocomplete_for_member.rhtml
	app/views/members/_member_form_impaired.rhtml

Conflicts:
	app/controllers/members_controller.rb
	app/views/layouts/base.rhtml
	public/javascripts/application.js

Conflicts:
	app/assets/images/suche_lupe.png
	app/controllers/members_controller.rb
	app/views/members/_autocomplete_for_member.rhtml
	app/views/projects/settings/_members.rhtml
	public/images/search_magnifier.png
	public/images/suche_lupe.png
	public/stylesheets/application.css
	public/stylesheets/select2_customizing.css

Conflicts:
	config/locales/de.yml
	config/locales/en.yml

GarstgerUnhold committed:
readd role search translation

GarstgerUnhold committed:
pluralize search translations, as we use multi-selects

nkenneweg committed:
we need an id for find_project not a project_id.

nkenneweg commented on app/models/principal.rb:
```diff
@@ -27,8 +27,13 @@ class Principal < ActiveRecord::Base
   scope :like, lambda {|q|
+    concatenation = "((firstname || ' ') || lastname)"
+    # special concat for mysql
+    if ChiliProject::Database.mysql?
```
Do we really need special Mysql code here?
Database specific code is always a bad thing, isn't it?

GarstgerUnhold replied:
Yes, it's definitely bad. But maybe it's just that it's friday, but I don't see a clean solution around this. I mean, you don't have access to the actual object since it's a scope and that would make no sense so we can't just concat strings on the ruby-side. And sadly you cannot use scopes inside of conditionals at least I'm not aware of that. So we're stuck with a helper I guess? Maybe ditch the scope in favor of plain ol' Arel, like so http://www.railway.at/2010/03/09/named-scopes-are-dead/? Am I missing something?

nkenneweg commented:
The CI still reports errors.

GarstgerUnhold committed:
fixes issue with mass_assignment protection and role_ids

GarstgerUnhold committed:
fixes membership creation for a single user

GarstgerUnhold committed:
reformatting the new member form inside project settings

Conflicts:

	app/views/members/_member_form.rhtml
	app/views/projects/settings/_members.rhtml
	public/stylesheets/application.css

Conflicts:

	app/views/projects/settings/_members.rhtml

Conflicts:
	app/views/projects/settings/_members.rhtml
	public/stylesheets/application.css

Conflicts:
	app/views/projects/settings/_members.html.erb

Conflicts:
	app/views/projects/settings/_members.html.erb

Conflicts:
	app/views/projects/settings/_members.html.erb

GarstgerUnhold committed:
Reworked Tests:

* fixes broken existing principal and group membership features
* adds new tests for adding groups and users as principals with both an impaired and non-impaired user

Conflicts:
	features/groups/group_memberships.feature
	features/step_definitions/project_member_steps.rb

sschu commented:
Travis says everything is fine and functionality looks good ... :thumbsup:
